src/clr-angular/utils/drag-and-drop/drag-and-drop-integrated.spec.ts

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Component, ElementRef } from '@angular/core';
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { By } from '@angular/platform-browser';
+import { NoopAnimationsModule } from '@angular/platform-browser/animations';
+
+import { ClrDragAndDropModule } from './drag-and-drop.module';
+import { ClrDraggable } from './draggable/draggable';
+import { emulateDragEvent } from './helpers.spec';
+
+// Here, we test drag and drop directives and components altogether in an integrated way.
+// Unlike other .spec test files, we use actual events in the following tests.
+
+const MOCK_DATA_PAYLOAD = {
+  data: 'test_drag_data_transfer',
+};
+
+export default function(): void {
+  describe('With ClrDraggable', function() {
+    let fixture: ComponentFixture<any>;
+    let testComponent: WithDraggableTest;
+    let testElement: HTMLElement;
+    let draggable: ElementRef;
+
+    beforeEach(function() {
+      TestBed.configureTestingModule({
+        imports: [ClrDragAndDropModule, NoopAnimationsModule],
+        declarations: [WithDraggableTest],
+      });
+
+      fixture = TestBed.createComponent(WithDraggableTest);
+      testComponent = fixture.componentInstance;
+      testElement = fixture.nativeElement;
+
+      draggable = fixture.debugElement.queryAll(By.directive(ClrDraggable))[1];
+      fixture.detectChanges();
+    });
+
+    afterEach(function() {
+      fixture.destroy();
+    });
+
+    describe('Template API', function() {
+      const checkStaticProps = function(dragEvent: any) {
+        // the following properties don't change accross different drag events
+        // and expected to stay the same in all of them.
+        expect(dragEvent.group).toBe('draggable-1');
+        expect(dragEvent.dragDataTransfer).toEqual(MOCK_DATA_PAYLOAD);
+      };
+
+      it('should emit event with proper properties on dragStart', function() {
+        emulateDragEvent('mousedown', 0, 0, draggable.nativeElement);
+        emulateDragEvent('mousemove', 10, 20);
+        checkStaticProps(testComponent.dragStartEvent);
+        expect(testComponent.dragStartEvent.dragPosition).toEqual({ pageX: 10, pageY: 20 });
+        expect(testComponent.dragStartEvent.dropPointPosition).toBeUndefined();
+      });
+
+      it('should emit event with proper properties on dragMove', function() {
+        emulateDragEvent('mousedown', 0, 0, draggable.nativeElement);
+        emulateDragEvent('mousemove', 10, 20);
+        emulateDragEvent('mousemove', 100, 200);
+        checkStaticProps(testComponent.dragMoveEvent);
+        expect(testComponent.dragMoveEvent.dragPosition).toEqual({ pageX: 100, pageY: 200 });
+
+        // offsetX = dragStart.pageX - draggable.pageX
```

Comment:
I'm not sure how this will pass if you run the same expect twice without changing properties.

Reply:
Sorry, would you elaborate more on this? Which line are you talking about?

Reply:
Oh I guess the two expects are for different points. However, do you need these commented out?

Reply:
That's basically the formula of how `dropPointPosition` was being calculated. Just to show or remind people.